### ISSUE TITLE
embed through editor adds huge padding-bottom buffer on each embed

### ISSUE THREAD

**Issue by emullaney** [wagtail/wagtail#1495]
Embeds added through the hallo editor seem to wrap iframes inside of the following div with an inline style:

``` html
<div style="padding-bottom: 56.25%;" class="responsive-object">
 iframe
</div>
```

I had to override this in my project css with the following code:

``` css
.responsive-object {
  padding-bottom: 0px !important;
}
```

**Comment by timheap** [comment]
As best I can figure, the padding is inserted with the assumption that you include some other styles that then make the whole embed responsive. I've not actually found these styles in the documentation, but I have made some styles myself that seem to work. Try using these styles instead:

```css
.responsive-object {
    position: relative;
}
.responsive-object iframe,
.responsive-object object {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
}
```

**Comment by davecranwell** [comment]
Correct, that padding is to automatically make any embedded object responsive in height and width, but it does assume the existence of some CSS.

It's documented here: http://docs.wagtail.io/en/v0.8.8/core_components/pages/writing_templates.html?highlight=responsive#responsive-embeds

**Comment by emullaney** [comment]
Thanks much for posting this, Dave! 

Will this documentation make it into the updated docs for the latest versions of wagtail?

**Comment by davecranwell** [comment]
It's moved around a little but yes, it's already there: http://docs.wagtail.io/en/latest/topics/writing_templates.html?highlight=responsive#responsive-embeds

**Comment by emullaney** [comment]
Ah, cool. I couldn't find it with the search functionality in the docs, when I was searching, which led me to open this issue in the first place. Searching for "responsive-object" doesn't bring up any results:

http://docs.wagtail.io/en/latest/search.html?q=responsive-object&check_keywords=yes&area=default

Anyway... thanks again, this was very helpful. We're launching next week!

**Comment by gasman** [comment]
Reopening this issue, as this has caught plenty of people out, and relying on a snippet of CSS buried in the docs isn't really sufficient. Embeds should be made to work sensibly out of the box, either with or without responsive behaviour baked in.

**Comment by davecranwell** [comment]
I think the responsive behaviour should be removed. It's not like there's any shortage of example code or plugins out there to achieve it.

**Comment by gasman** [comment]
Sounds good to me. We'd need to come up with an upgrade path for people who are using the current responsive behaviour, though. I see that the current implementation relies on some server-side logic to calculate the embed's aspect ratio, which is why it has to be an inline style... @davecranwell, do the plugins/code examples you mention avoid the need for this server-side code?

If the server-side logic can't be avoided (or even if it can...), then one neat way of dealing with it would be to move the responsive code to a contrib module, and tell people 'to restore the old behaviour, add `wagtail.contrib.responsive_embeds` to INSTALLED_APPS'. (It looks like this ought to be possible with our existing richtext hooks, but no big deal if we need to add new ones.)

**Comment by davecranwell** [comment]
Most of the plugins I've seen calculate the aspect ratio of an embed from the width/height or attribute properties of the elements provided by the third party. e.g the Youtube `iframe's `width` and `height` props.

Not only does this mean the developer has to implement additional plugins themselves, but often they have to implement the same code umpteen times to account for different embedded media types. Having Wagtail do this for you, for _anything_, is definitely very nice in that respect.

Ideally I'd like to still provide this ratio information to developer, but without it actually taking effect on the elements. Perhaps that's not possible though.

**Comment by gasman** [comment]
Sounds like the contrib app is the way to go then. It would be nice if we could get that app to add the rest of the CSS as well, but that's probably best left for the future as a big-picture piece of development (i.e. "a standardised way for Wagtail pluggable apps to insert resources into the site front-end").

**Comment by AtomicSpark** [comment]
I second the optional contrib suggestion. Wagtail, by default, should not force any CSS into a web page. The reason that I chose Wagtail was because it seemed to best follow Django's "clean, pragmatic design" and ~~forced~~ allowed me to write my HTML templates from scratch.

I'm adding the CSS snippet from the Wagtail documentation to fix the issue, but I'm not happy about it.

**Comment by davecranwell** [comment]
Having discussed this with the team, the approach will be as suggested above. We move the current responsive behaviour into a new app, then remove all remaining responsiveness or dependency on CSS, from the embed system. 

Embeds will also be better templated with an {% embed %} template tag, similar to the {% image %} which will provide access to deeper variables about an embed. Included in these "deeper variables" will be the ratio of the video which _will_ continue to be calculated by Wagtail's internals. Front End developers wanting responsive embeds will be able to use these ratio information provided by the {% embed %} tag, to facilitate whatever CSS/JS tricky they choose to employ.

**Comment by ghostwords** [comment]
Adding @timheap's CSS fixes the padding following YouTube embeds (and makes them responsive), but makes Instagram embeds disappear. You should be able to reproduce on the Wagtail demo site using the default oEmbed provider.

**Comment by gasman** [comment]
@ghostwords Does this happen if you try the CSS from http://docs.wagtail.io/en/v0.8.8/core_components/pages/writing_templates.html#responsive-embeds ?

**Comment by ghostwords** [comment]
@gasman Yep ...

Without the CSS (previewing a YouTube video and an Instagram embed):
![wagtail_embed_without_css](https://cloud.githubusercontent.com/assets/794578/11016171/e160039e-8546-11e5-9c71-a333c4e225ec.png)

Note the padding after the video.

Same page, with the CSS (YouTube is good, Instagram is gone):
![wagtail_embed_with_css](https://cloud.githubusercontent.com/assets/794578/11016176/e5fa056c-8546-11e5-800a-c43a960133c5.png)

**Comment by SalahAdDin** [comment]
:+1:

**Comment by Gagaro** [comment]
I also have the issue. Is there any news about moving the responsive part to a contrib app?

The problem with instagram is also still present. A workaround I found is to disable the responsive behavior if no size is provided.

```django
<div{% if ratio != "0" %} style="padding-bottom: {{ ratio }};" class="responsive-object"{% endif %}>
    {{ embed.html|safe }}
</div>
```

I'm working on the PR right now for this small fix.

**Comment by coredumperror** [comment]
Is the move of responsive embeds to contrib still in progress? I ran into this problem with the spurious `padding-bottom` rule yesterday, and lost a lot of time making it go away. I wish I'd known about that obscure bit of CSS earlier...

**Comment by Gagaro** [comment]
It should have been fixed in 1d33cc7838c647c95f4c228b35500851cbd5e6d9. You have the padding if you have a ratio available (if you have the size of the embed). What was your issue exactly?

**Comment by coredumperror** [comment]
My issue was that embedding a video into my page caused the div the video was in to be almost 60% taller than it was supposed to be, due to that `padding-bottom` style being added in the markup. I didn't want that extra padding, so I did a bunch of work to override the code that generated it. Then I discovered that bit of CSS in the docs, which makes the video's div look normal. So I undid all those code overrides, and everything seems to be fine, now.

**Comment by lb-** [comment]
Just flagging that this came up again in a recent question on Stack Overflow (question now deleted), so it is still catching people out.

Most recent docs link to the solution: http://docs.wagtail.io/en/v1.13.1/topics/writing_templates.html#responsive-embeds

**Comment by thenewguy** [comment]
It would be helpful if the css was bundled with wagtail so one could just include it at the top of their template

```
<link rel="stylesheet" href="{% static 'wagtail/css/some-name-like-responive-embeds.css' %}" />
```

**Comment by hanzvanaardt** [comment]
I ran into this issue today and it sucked in about four hours of trying to figure out what is going on.  I was referred to this thread.  Creating an HTML element with an embedded style cannot be the correct solution can it?  This HTML gets created when embedding a Youtube video using the Rich Text editor <div style="padding-bottom: 56.25%;" class="responsive-object"> - the only way I can get rid of this style is to override it in my CSS using !important which is terrible and basically freezes any future changes to this style.

**Comment by tanero** [comment]
I was trying to understand what was the problem for embedding a video, but it was a basic css bug comes with wagtail.  But it is not clear will be fixed on new version or not.

**Comment by autoferrit** [comment]
This still occurs for me too in the rich text editor. I just clicked the "embed video" button and pasted the video.

**Comment by gasman** [comment]
Fixed in #5769
